File: backend/src/services/CleanupService.ts
import cron from 'node-cron';
import { prisma } from '../db/prisma';
import * as StorageService from './StorageService';

/**
 * CleanupService — runs every 5 minutes to purge expired shares.
 * 1. Finds all shares past their expiresAt timestamp
 * 2. Deletes associated files from R2
 * 3. Deletes DB records (cascades to File + TextShare)
 */

let isRunning = false;

async function cleanupExpiredShares(): Promise<void> {
  // Prevent overlapping runs
  if (isRunning) return;
  isRunning = true;

  try {
    const expired = await prisma.share.findMany({
      where: {
        expiresAt: {
          lte: new Date(),
          not: null,
        },
      },
      include: {
        files: { select: { storageKey: true } },
      },
    });

    if (expired.length === 0) {
      isRunning = false;
      return;
    }

    console.log(`🧹 Cleaning up ${expired.length} expired share(s)...`);

    for (const share of expired) {
      try {
        // Delete files from R2
        if (share.files.length > 0) {
          const keys = share.files.map((f) => f.storageKey);
          await StorageService.deleteFiles(keys);
        }

        // Delete DB record (cascades to files + textShare)
        await prisma.share.delete({ where: { id: share.id } });

        console.log(`  ✓ Deleted share ${share.uid}`);
      } catch (error) {
        // Log but continue — next run will retry failed deletions
        console.error(`  ✗ Failed to clean up share ${share.uid}:`, error);
      }
    }
  } catch (error) {
    console.error('❌ Cleanup job failed:', error);
  } finally {
    isRunning = false;
  }
}

/**
 * Start the cleanup cron job (every 5 minutes).
 */
export function startCleanupWorker(): void {
  console.log('🕐 Cleanup worker started — running every 5 minutes');

  // Run immediately on startup to catch any missed expiries
  cleanupExpiredShares();

  // Schedule recurring job
  cron.schedule('*/5 * * * *', cleanupExpiredShares);
}

File: backend/app/services/cleanup_service.py
"""
cleanup_service.py — APScheduler-based cleanup worker.
Replaces: backend/src/services/CleanupService.ts (node-cron)
"""

import asyncio
import logging
from datetime import datetime, timezone

from sqlalchemy import select, delete
from sqlalchemy.orm import selectinload

from app.database import async_session
from app.models import Share
from app.services import storage_service

logger = logging.getLogger("sharenova.cleanup")

_running = False


async def cleanup_expired_shares() -> None:
    """Delete expired shares from DB and R2. Overlap-prevention via _running flag."""
    global _running
    if _running:
        logger.debug("Cleanup already running — skipping")
        return

    _running = True
    try:
        async with async_session() as db:
            # Find expired shares
            result = await db.execute(
                select(Share)
                .where(Share.expires_at <= datetime.now(timezone.utc))
                .where(Share.expires_at.isnot(None))
                .options(selectinload(Share.files))
            )
            expired = result.scalars().all()

            if not expired:
                return

            # Collect storage keys and delete from R2
            storage_keys = []
            for share in expired:
                for f in share.files:
                    storage_keys.append(f.storage_key)

            if storage_keys:
                try:
                    await storage_service.delete_files(storage_keys)
                except Exception as e:
                    logger.error(f"Failed to delete files from R2: {e}")

            # Delete from DB (cascade handles files + text_shares)
            share_ids = [s.id for s in expired]
            await db.execute(delete(Share).where(Share.id.in_(share_ids)))
            await db.commit()

            logger.info(f"Cleaned up {len(expired)} expired shares, {len(storage_keys)} files")
    except Exception as e:
        logger.error(f"Cleanup error: {e}")
    finally:
        _running = False
